compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs

```diff
@@ -72,21 +68,29 @@ pub fn validate_cmse_abi<'tcx>(
 fn is_valid_cmse_inputs<'tcx>(
     tcx: TyCtxt<'tcx>,
     fn_sig: ty::PolyFnSig<'tcx>,
-) -> Result<bool, &'tcx LayoutError<'tcx>> {
+) -> Result<Result<(), usize>, &'tcx LayoutError<'tcx>> {
+    let mut span = None;
     let mut accum = 0u64;
 
-    for arg_def in fn_sig.inputs().iter() {
+    for (index, arg_def) in fn_sig.inputs().iter().enumerate() {
         let layout = tcx.layout_of(ParamEnv::reveal_all().and(*arg_def.skip_binder()))?;
```

Comment:
I think at this point it may be easier to just emit the errors in `is_valid_cmse_inputs` and `_output` directly. Then you can also report a "generics not allowed" error with a span pointing at the offending thing.

Reply:
we have 2 reasons not to do that

* `arg_def` doesn't have any span information. it's effectively just a `Ty`
* in the (near, hopefully) future, the `C-cmse-nonsecure-entry` abi will require the same validation, and we'd like to reuse this code, so passing in `BareFn`-specific data will not work